File: DeepThink/Services/EmbeddingService.swift
import Foundation
import NaturalLanguage
import SwiftData

@Observable
final class EmbeddingService {
    static let shared = EmbeddingService()

    private(set) var isIndexing = false
    private(set) var indexedCount = 0
    private(set) var progress: Double = 0

    private let nlEmbedding: NLEmbedding?
    private let store = VectorStore.shared
    private let embedQueue = DispatchQueue(label: "com.deepthink.embedding.nlp")
    private let drainQueue = DispatchQueue(label: "com.deepthink.embedding.drain", qos: .utility)
    private var reconcilerTask: Task<Void, Never>?

    private init() {
        nlEmbedding = NLEmbedding.sentenceEmbedding(for: .english)
        indexedCount = store.embeddedCount()
    }

    // MARK: - Embedding

    func embed(_ text: String) -> [Double]? {
        guard let nlEmbedding else { return nil }
        return embedQueue.sync {
            guard let vector = nlEmbedding.vector(for: text) else { return nil }
            guard !vector.contains(where: { $0.isNaN || $0.isInfinite }) else { return nil }
            return vector
        }
    }

    // MARK: - Knowledge Indexing

    func indexEntries(_ entries: [KnowledgeEntry]) {
        DispatchQueue.main.async { self.isIndexing = true }
        defer {
            let count = store.embeddedCount()
            DispatchQueue.main.async { self.isIndexing = false; self.indexedCount = count }
        }

        let total = entries.count
        for (i, entry) in entries.enumerated() {
            indexSingleEntry(entry)
            let p = Double(i + 1) / Double(total)
            DispatchQueue.main.async { self.progress = p }
        }

        store.pruneStaleEntries(
            validIDs: Set(entries.map(\.id)),
            entryType: "knowledge"
        )
    }

    func indexSingleEntry(_ entry: KnowledgeEntry) {
        let hash = simpleHash(entry.content)
        if let existing = store.contentHash(forEntry: entry.id), existing == hash {
            return
        }

        let chunks = SemanticChunker.chunk(
            text: entry.content,
            entryID: entry.id,
            entryType: "knowledge",
            title: entry.title,
            tags: entry.tags,
            source: entry.source,
            importedAt: entry.importedAt,
            contentHash: hash
        )

        let vectorChunks = chunks.map { chunk -> VectorChunk in
            let text = "\(entry.title). \(String(chunk.content.prefix(500)))"
            let embedding = embed(text)
            return VectorChunk(
                id: chunk.id, entryID: chunk.entryID, entryType: chunk.entryType,
                title: chunk.title, content: chunk.content, tags: chunk.tags,
                source: chunk.source, importedAt: chunk.importedAt,
                chunkIndex: chunk.chunkIndex, totalChunks: chunk.totalChunks,
                contentHash: chunk.contentHash, embedding: embedding
            )
        }

        store.replaceChunksForEntry(entry.id, with: vectorChunks)
    }

    func removeEntry(_ entryID: String) {
        store.deleteChunksForEntry(entryID)
        indexedCount = store.embeddedCount()
    }

    // MARK: - Workspace Indexing (Tasks, Notes, Reminders)

    func indexWorkspaceItem(
        id: String,
        type: String,
        title: String,
        content: String,
        tags: [String] = [],
        modifiedAt: Date = Date()
    ) throws {
        let hash = simpleHash(content)
        if let existing = store.contentHash(forEntry: id), existing == hash {
            return
        }

        let fullText = "\(title). \(content)"
        let chunks = SemanticChunker.chunk(
            text: fullText.count > 200 ? content : fullText,
            entryID: id,
            entryType: type,
            title: title,
            tags: tags,
            source: type,
            importedAt: modifiedAt,
            contentHash: hash
        )

        let vectorChunks = chunks.map { chunk -> VectorChunk in
            let embedding = embed("\(title). \(String(chunk.content.prefix(500)))")
            return VectorChunk(
                id: chunk.id, entryID: chunk.entryID, entryType: chunk.entryType,
                title: chunk.title, content: chunk.content, tags: chunk.tags,
                source: chunk.source, importedAt: chunk.importedAt,
                chunkIndex: chunk.chunkIndex, totalChunks: chunk.totalChunks,
                contentHash: chunk.contentHash, embedding: embedding
            )
        }

        if vectorChunks.allSatisfy({ $0.embedding == nil }) {
            throw EmbeddingError.allChunksFailed
        }

        store.replaceChunksForEntry(id, with: vectorChunks)
    }

    func indexWorkspaceItems(_ items: [(id: String, type: String, title: String, content: String, tags: [String], modifiedAt: Date)]) {
        DispatchQueue.main.async { self.isIndexing = true }
        defer {
            let count = store.embeddedCount()
            DispatchQueue.main.async { self.isIndexing = false; self.indexedCount = count }
        }

        let total = items.count
        for (i, item) in items.enumerated() {
            do {
                try indexWorkspaceItem(
                    id: item.id, type: item.type, title: item.title,
                    content: item.content, tags: item.tags, modifiedAt: item.modifiedAt
                )
            } catch {
                VectorStore.shared.enqueuePendingReindex(entryID: item.id, entryType: item.type)
            }
            let p = Double(i + 1) / Double(total)
            DispatchQueue.main.async { self.progress = p }
        }
    }

    // MARK: - Pending Reindex Drain

    func drainPendingReindex(container: ModelContainer) {
        drainQueue.async {
            let pending = VectorStore.shared.fetchPendingReindex(maxRetries: 3)
            guard !pending.isEmpty else { return }

            let context = ModelContext(container)

            for row in pending {
                do {
                    if row.operation == "delete" {
                        VectorStore.shared.deleteChunksForEntry(row.entryID)
                        VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                        continue
                    }

                    let parts = row.entryID.split(separator: ":", maxSplits: 1)
                    guard parts.count == 2, let uuid = UUID(uuidString: String(parts[1])) else {
                        VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                        continue
                    }

                    switch row.entryType {
                    case "task":
                        let tasks = (try? context.fetch(FetchDescriptor<TaskItem>(predicate: #Predicate { $0.id == uuid }))) ?? []
                        guard let task = tasks.first else {
                            // Item deleted — clean up any orphaned chunks
                            VectorStore.shared.deleteChunksForEntry(row.entryID)
                            VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                            continue
                        }
                        try self.indexWorkspaceItem(
                            id: row.entryID, type: "task",
                            title: task.title,
                            content: Self.taskContent(title: task.title, detail: task.detail, status: task.statusRaw, isArchived: task.isArchived),
                            tags: task.tags.map(\.name), modifiedAt: task.modifiedAt
                        )
                    case "note":
                        let notes = (try? context.fetch(FetchDescriptor<Note>(predicate: #Predicate { $0.id == uuid }))) ?? []
                        guard let note = notes.first else {
                            VectorStore.shared.deleteChunksForEntry(row.entryID)
                            VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                            continue
                        }
                        try self.indexWorkspaceItem(
                            id: row.entryID, type: "note",
                            title: note.title,
                            content: Self.noteContent(title: note.title, content: note.content, isArchived: note.isArchived),
                            tags: note.tags.map(\.name), modifiedAt: note.modifiedAt
                        )
                    case "reminder":
                        let reminders = (try? context.fetch(FetchDescriptor<Reminder>(predicate: #Predicate { $0.id == uuid }))) ?? []
                        guard let reminder = reminders.first else {
                            VectorStore.shared.deleteChunksForEntry(row.entryID)
                            VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                            continue
                        }
                        try self.indexWorkspaceItem(
                            id: row.entryID, type: "reminder",
                            title: reminder.title,
                            content: Self.reminderContent(title: reminder.title, notes: reminder.notes, isCompleted: reminder.isCompleted),
                            modifiedAt: reminder.modifiedAt
                        )
                    default:
                        VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                        continue
                    }
                    VectorStore.shared.deletePendingReindex(entryID: row.entryID)
                } catch {
                    VectorStore.shared.incrementPendingRetry(entryID: row.entryID)
                }
            }

            let count = VectorStore.shared.embeddedCount()
            DispatchQueue.main.async { self.indexedCount = count }
        }
    }

    func startReconcilerTimer(container: ModelContainer) {
        reconcilerTask?.cancel()
        reconcilerTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(60))
                guard !Task.isCancelled else { break }
                drainPendingReindex(container: container)
            }
        }
    }

    // MARK: - Content String Builders (canonical — keep in sync with TypeScript side)

    static func taskContent(title: String, detail: String, status: String, isArchived: Bool) -> String {
        "\(title)\n\(detail)\nstatus:\(status)\narchived:\(isArchived)"
    }

    static func noteContent(title: String, content: String, isArchived: Bool) -> String {
        "\(title)\n\(content)\narchived:\(isArchived)"
    }

    static func reminderContent(title: String, notes: String, isCompleted: Bool) -> String {
        "\(title)\n\(notes)\ncompleted:\(isCompleted)"
    }

    // MARK: - Search

    func search(query: String, topK: Int = 10, scope: [String]? = nil, entryType: String? = nil) -> [(entryID: String, score: Double)] {
        guard let queryVector = embed(query) else { return [] }

        let entries = store.chunksWithEmbeddings(entryType: entryType, scope: scope)
        var bestScores: [String: Double] = [:]

        for (chunk, embedding) in entries {
            let similarity = cosineSimilarity(queryVector, embedding)
            if similarity > 0.3 {
                let current = bestScores[chunk.entryID] ?? 0
                if similarity > current {
                    bestScores[chunk.entryID] = similarity
                }
            }
        }

        return bestScores
            .map { (entryID: $0.key, score: $0.value) }
            .sorted { $0.score > $1.score }
            .prefix(topK)
            .map(\.self)
    }

    // MARK: - Stats

    var stats: (total: Int, embedded: Int, knowledge: Int, tasks: Int, notes: Int, reminders: Int) {
        (
            total: store.chunkCount(),
            embedded: store.embeddedCount(),
            knowledge: store.entryCount(entryType: "knowledge"),
            tasks: store.entryCount(entryType: "task"),
            notes: store.entryCount(entryType: "note"),
            reminders: store.entryCount(entryType: "reminder")
        )
    }

    // MARK: - Private

    private func cosineSimilarity(_ a: [Double], _ b: [Double]) -> Double {
        guard a.count == b.count, !a.isEmpty else { return 0 }
        var dot = 0.0, normA = 0.0, normB = 0.0
        for i in 0..<a.count {
            dot += a[i] * b[i]
            normA += a[i] * a[i]
            normB += b[i] * b[i]
        }
        let denom = sqrt(normA) * sqrt(normB)
        return denom > 0 ? dot / denom : 0
    }

    // 32-bit djb2 addition over UTF-8 bytes — matches TypeScript's simpleHash exactly.
    private func simpleHash(_ text: String) -> UInt64 {
        var hash: UInt32 = 5381
        for byte in text.utf8 {
            hash = hash &* 33 &+ UInt32(byte)
        }
        return UInt64(hash)
    }
}

// MARK: - Errors

enum EmbeddingError: Error {
    case allChunksFailed
}

// MARK: - Semantic Chunker

enum SemanticChunker {
    static let maxChunkSize = 500
    static let minChunkSize = 100

    static func chunk(
        text: String,
        entryID: String,
        entryType: String,
        title: String,
        tags: [String],
        source: String,
        importedAt: Date,
        contentHash: UInt64
    ) -> [VectorChunk] {
        let sentences = splitSentences(text)
        guard !sentences.isEmpty else {
            return [VectorChunk(
                id: "\(entryID):0", entryID: entryID, entryType: entryType,
                title: title, content: text, tags: tags, source: source,
                importedAt: importedAt, chunkIndex: 0, totalChunks: 1,
                contentHash: contentHash, embedding: nil
            )]
        }

        var groups: [[String]] = []
        var current: [String] = []
        var currentLen = 0

        for sentence in sentences {
            if currentLen + sentence.count > maxChunkSize, !current.isEmpty {
                groups.append(current)
                // Overlap: keep last sentence
                let last = current.last ?? ""
                current = last.count < maxChunkSize / 2 ? [last] : []
                currentLen = current.reduce(0) { $0 + $1.count }
            }
            current.append(sentence)
            currentLen += sentence.count
        }

        if !current.isEmpty {
            if currentLen < minChunkSize, let last = groups.last {
                groups[groups.count - 1] = last + current
            } else {
                groups.append(current)
            }
        }

        let totalChunks = groups.count
        return groups.enumerated().map { index, group in
            VectorChunk(
                id: "\(entryID):\(index)",
                entryID: entryID, entryType: entryType,
                title: title, content: group.joined(separator: " "),
                tags: tags, source: source, importedAt: importedAt,
                chunkIndex: index, totalChunks: totalChunks,
                contentHash: contentHash, embedding: nil
            )
        }
    }

    private static func splitSentences(_ text: String) -> [String] {
        var sentences: [String] = []
        text.enumerateSubstrings(in: text.startIndex..., options: [.bySentences, .localized]) { substring, _, _, _ in
            if let s = substring?.trimmingCharacters(in: .whitespacesAndNewlines), !s.isEmpty {
                sentences.append(s)
            }
        }
        if sentences.isEmpty, !text.isEmpty {
            return text.components(separatedBy: "\n")
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
        }
        return sentences
    }
}

File: cli/src/core/embedding-service.ts
import { execFileSync, execSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { DEEPTHINK_ROOT } from "../config";
import * as db from "./db";
import { hexToUUID } from "./db";
import {
  batchContentHashes,
  chunksWithEmbeddings,
  deleteChunksForEntry,
  deleteExhaustedPendingReindex,
  deletePendingReindex,
  embeddedCount,
  enqueuePendingReindex,
  contentHash as getContentHash,
  getPendingReindex,
  incrementPendingRetry,
  pruneStaleEntries,
  replaceChunksForEntry,
  semanticChunk,
  simpleHash,
} from "./vector-store";

const CACHE_DIR = join(DEEPTHINK_ROOT, ".cache");
const HELPER_BIN = join(CACHE_DIR, "embed-helper");
const HELPER_SRC = join(CACHE_DIR, "embed-helper.swift");
const HELPER_VERSION_FILE = join(CACHE_DIR, "embed-helper.version");

// Bump this whenever SWIFT_SOURCE changes so stale binaries are auto-recompiled.
const SWIFT_VERSION = "2";

const _embeddingCache = new Map<string, { vector: Float32Array; ts: number }>();
const EMBEDDING_CACHE_TTL = 5 * 60 * 1000;
const EMBEDDING_CACHE_MAX = 200;

const SWIFT_SOURCE = `import NaturalLanguage
import Foundation

let args = CommandLine.arguments.dropFirst()

if args.first == "--batch" {
    // Batch mode: read one text per line from stdin, emit one CSV vector per line.
    // An empty output line signals embedding failure for that input.
    guard let embedding = NLEmbedding.sentenceEmbedding(for: .english) else {
        fputs("error: NLEmbedding unavailable\\n", stderr)
        exit(1)
    }
    while let line = readLine(strippingNewline: true) {
        if let vector = embedding.vector(for: line) {
            print(vector.map { String($0) }.joined(separator: ","))
        } else {
            print("")
        }
    }
} else {
    // Single-arg mode (backward compat)
    guard !args.isEmpty else {
        fputs("usage: embed-helper <text>\\n", stderr)
        exit(1)
    }
    let text = args.joined(separator: " ")
    guard let embedding = NLEmbedding.sentenceEmbedding(for: .english),
          let vector = embedding.vector(for: text) else {
        fputs("error: NLEmbedding unavailable\\n", stderr)
        exit(1)
    }
    print(vector.map { String($0) }.joined(separator: ","))
}
`;

function ensureHelper(): boolean {
  if (existsSync(HELPER_BIN) && existsSync(HELPER_VERSION_FILE)) {
    try {
      if (readFileSync(HELPER_VERSION_FILE, "utf-8").trim() === SWIFT_VERSION) return true;
    } catch {}
    // Version mismatch — delete stale artifacts and recompile.
    try {
      unlinkSync(HELPER_BIN);
    } catch {}
    try {
      unlinkSync(HELPER_VERSION_FILE);
    } catch {}
  } else if (existsSync(HELPER_BIN)) {
    // Binary exists but no version file — treat as stale.
    try {
      unlinkSync(HELPER_BIN);
    } catch {}
  }
  try {
    if (!existsSync(CACHE_DIR)) mkdirSync(CACHE_DIR, { recursive: true });
    writeFileSync(HELPER_SRC, SWIFT_SOURCE);
    execSync(`swiftc -O -o '${HELPER_BIN}' '${HELPER_SRC}'`, {
      timeout: 30000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    writeFileSync(HELPER_VERSION_FILE, SWIFT_VERSION);
    return true;
  } catch {
    return false;
  }
}

// Embeds multiple texts in a single subprocess call. Cache is consulted/updated per text.
function embedBatch(texts: string[]): (Float32Array | null)[] {
  if (texts.length === 0) return [];
  if (!ensureHelper()) return texts.map(() => null);

  const now = Date.now();
  const results: (Float32Array | null)[] = new Array(texts.length).fill(null);
  const uncachedIndices: number[] = [];
  const uncachedTexts: string[] = [];

  for (let i = 0; i < texts.length; i++) {
    const cached = _embeddingCache.get(texts[i]);
    if (cached && now - cached.ts < EMBEDDING_CACHE_TTL) {
      results[i] = cached.vector;
    } else {
      uncachedIndices.push(i);
      uncachedTexts.push(texts[i]);
    }
  }

  if (uncachedTexts.length === 0) return results;

  try {
    // Newlines would break the line-delimited protocol — replace with spaces.
    const stdin = uncachedTexts.map((t) => t.replace(/\n/g, " ")).join("\n");
    const output = execFileSync(HELPER_BIN, ["--batch"], {
      input: stdin,
      timeout: Math.max(10000, uncachedTexts.length * 800),
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    const lines = output.split("\n");
    for (let j = 0; j < uncachedIndices.length; j++) {
      const line = (lines[j] ?? "").trim();
      if (!line) continue;
      const values = line.split(",").map(Number);
      if (values.length > 0 && !values.some(Number.isNaN)) {
        const vector = new Float32Array(values);
        if (_embeddingCache.size >= EMBEDDING_CACHE_MAX) {
          _embeddingCache.delete(_embeddingCache.keys().next().value!);
        }
        _embeddingCache.set(texts[uncachedIndices[j]], { vector, ts: now });
        results[uncachedIndices[j]] = vector;
      }
    }
  } catch {
    // Leave nulls for uncached texts on failure.
  }

  return results;
}

function embedQuery(text: string): Float32Array | null {
  return embedBatch([text])[0] ?? null;
}

function cosineSimilarity(a: number[] | Float32Array, b: number[] | Float32Array): number {
  if (a.length !== b.length || a.length === 0) return 0;
  let dot = 0,
    normA = 0,
    normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom > 0 ? dot / denom : 0;
}

// MARK: - Indexing

export interface IndexableEntry {
  id: string;
  type: string;
  title: string;
  content: string;
  tags: string[];
  source: string;
  importedAt: Date;
}

// Internal: accepts a pre-loaded hash map to avoid per-entry DB queries.
function indexEntryCore(entry: IndexableEntry, knownHashes?: Map<string, number>): void {
  const hash = simpleHash(entry.content);
  const existing = knownHashes !== undefined ? (knownHashes.get(entry.id) ?? null) : getContentHash(entry.id);
  if (existing !== null && existing === hash) return;

  const chunks = semanticChunk(
    entry.content,
    entry.id,
    entry.type,
    entry.title,
    entry.tags,
    entry.source,
    entry.importedAt,
    hash
  );

  const texts = chunks.map((c) => `${entry.title}. ${c.content.slice(0, 500)}`);
  const embeddings = embedBatch(texts);
  const withEmbeddings = chunks.map((chunk, i) => ({ ...chunk, embedding: embeddings[i] ?? null }));

  if (withEmbeddings.every((c) => c.embedding === null)) {
    throw new Error(`embedding failed for all chunks of ${entry.id} — will retry`);
  }

  replaceChunksForEntry(entry.id, withEmbeddings);
}

export function indexEntry(entry: IndexableEntry): void {
  try {
    indexEntryCore(entry);
  } catch {
    enqueuePendingReindex(entry.id, entry.type);
  }
}

export function indexEntries(entries: IndexableEntry[]): void {
  if (entries.length === 0) return;
  const knownHashes = batchContentHashes(entries.map((e) => e.id));
  for (const entry of entries) indexEntryCore(entry, knownHashes);
  const validIds = new Set(entries.map((e) => e.id));
  pruneStaleEntries(validIds, entries[0].type ?? "knowledge");
}

export function removeEntry(entryId: string): void {
  deleteChunksForEntry(entryId);
}

// Re-embed all workspace items that are missing or have stale embeddings.
// Safe to call multiple times — unchanged items are skipped.
export function reindexWorkspace(): { indexed: number } {
  const tasks = db.listTasks({ excludeArchived: false });
  const notes = db.listNotes({ excludeArchived: false });
  const reminders = db.listReminders({});
  const projects = db.listProjects();

  const entries: IndexableEntry[] = [
    ...tasks.map((t) => ({
      id: `task:${hexToUUID(t.id)}`,
      type: "task",
      title: t.title,
      content: taskContent(t.title, t.detail, t.status, t.isArchived),
      tags: [],
      source: t.isArchived ? "archive" : "task",
      importedAt: t.modifiedAt,
    })),
    ...notes.map((n) => ({
      id: `note:${hexToUUID(n.id)}`,
      type: "note",
      title: n.title,
      content: noteContent(n.title, n.content, n.isArchived),
      tags: [],
      source: n.isArchived ? "archive" : "note",
      importedAt: n.modifiedAt,
    })),
    ...reminders.map((r) => ({
      id: `reminder:${hexToUUID(r.id)}`,
      type: "reminder",
      title: r.title,
      content: reminderContent(r.title, r.notes, r.isCompleted),
      tags: [],
      source: "reminder",
      importedAt: r.modifiedAt,
    })),
    ...projects.map((p) => ({
      id: `project:${hexToUUID(p.id)}`,
      type: "project",
      title: p.name,
      content: projectContent(p.name, p.summary, p.isArchived),
      tags: [],
      source: p.isArchived ? "archive" : "project",
      importedAt: p.modifiedAt,
    })),
  ];

  const knownHashes = batchContentHashes(entries.map((e) => e.id));
  let indexed = 0;
  for (const entry of entries) {
    const hash = simpleHash(entry.content);
    if ((knownHashes.get(entry.id) ?? null) !== hash) {
      try {
        indexEntryCore(entry, knownHashes);
        indexed++;
      } catch {
        enqueuePendingReindex(entry.id, entry.type);
      }
    }
  }

  const validIds = new Set(entries.map((e) => e.id));
  pruneStaleEntries(new Set([...validIds].filter((id) => id.startsWith("task:"))), "task");
  pruneStaleEntries(new Set([...validIds].filter((id) => id.startsWith("note:"))), "note");
  pruneStaleEntries(new Set([...validIds].filter((id) => id.startsWith("reminder:"))), "reminder");
  pruneStaleEntries(new Set([...validIds].filter((id) => id.startsWith("project:"))), "project");

  return { indexed };
}

// Drain the pending_reindex queue — processes entries that failed inline indexing.
export function drainPendingReindex(): { processed: number; failed: number } {
  const pending = getPendingReindex(3);
  if (pending.length === 0) return { processed: 0, failed: 0 };

  const tasks = db.listTasks({ excludeArchived: false });
  const notes = db.listNotes({ excludeArchived: false });
  const reminders = db.listReminders({});
  const projects = db.listProjects();

  const taskMap = new Map(tasks.map((t) => [`task:${hexToUUID(t.id)}`, t]));
  const noteMap = new Map(notes.map((n) => [`note:${hexToUUID(n.id)}`, n]));
  const reminderMap = new Map(reminders.map((r) => [`reminder:${hexToUUID(r.id)}`, r]));
  const projectMap = new Map(projects.map((p) => [`project:${hexToUUID(p.id)}`, p]));

  let processed = 0;
  let failed = 0;

  for (const row of pending) {
    try {
      if (row.operation === "delete") {
        deleteChunksForEntry(row.entryId);
        deletePendingReindex(row.entryId);
        processed++;
        continue;
      }

      let entry: IndexableEntry | null = null;

      if (row.entryType === "task") {
        const t = taskMap.get(row.entryId);
        if (!t) { deletePendingReindex(row.entryId); continue; }
        entry = { id: row.entryId, type: "task", title: t.title, content: taskContent(t.title, t.detail, t.status, t.isArchived), tags: [], source: t.isArchived ? "archive" : "task", importedAt: t.modifiedAt };
      } else if (row.entryType === "note") {
        const n = noteMap.get(row.entryId);
        if (!n) { deletePendingReindex(row.entryId); continue; }
        entry = { id: row.entryId, type: "note", title: n.title, content: noteContent(n.title, n.content, n.isArchived), tags: [], source: n.isArchived ? "archive" : "note", importedAt: n.modifiedAt };
      } else if (row.entryType === "reminder") {
        const r = reminderMap.get(row.entryId);
        if (!r) { deletePendingReindex(row.entryId); continue; }
        entry = { id: row.entryId, type: "reminder", title: r.title, content: reminderContent(r.title, r.notes, r.isCompleted), tags: [], source: "reminder", importedAt: r.modifiedAt };
      } else if (row.entryType === "project") {
        const p = projectMap.get(row.entryId);
        if (!p) { deletePendingReindex(row.entryId); continue; }
        entry = { id: row.entryId, type: "project", title: p.name, content: projectContent(p.name, p.summary, p.isArchived), tags: [], source: p.isArchived ? "archive" : "project", importedAt: p.modifiedAt };
      }

      if (!entry) { deletePendingReindex(row.entryId); continue; }

      indexEntryCore(entry);
      deletePendingReindex(row.entryId);
      processed++;
    } catch {
      incrementPendingRetry(row.entryId);
      failed++;
    }
  }

  deleteExhaustedPendingReindex();
  return { processed, failed };
}

// Start a 30-second background reconciler that drains the pending_reindex queue.
export function startReconciler(): void {
  setInterval(() => { try { drainPendingReindex(); } catch {} }, 30_000);
}

// MARK: - Content string builders (canonical format — keep in sync with Swift side)

export function taskContent(title: string, detail: string, status: string, isArchived: boolean): string {
  return `${title}\n${detail}\nstatus:${status}\narchived:${isArchived}`;
}

export function noteContent(title: string, content: string, isArchived: boolean): string {
  return `${title}\n${content}\narchived:${isArchived}`;
}

export function reminderContent(title: string, notes: string | null | undefined, isCompleted: boolean): string {
  return `${title}\n${notes ?? ""}\ncompleted:${isCompleted}`;
}

export function projectContent(name: string, summary: string | null | undefined, isArchived: boolean): string {
  return `${name}\n${summary ?? ""}\narchived:${isArchived}`;
}

// MARK: - Search

export interface SemanticResult {
  entryID: string;
  score: number;
}

export function semanticSearch(query: string, topK: number = 10, scope?: string[]): SemanticResult[] {
  const queryVector = embedQuery(query);
  if (!queryVector) return [];

  const entries = chunksWithEmbeddings({ scope });
  // Track the best-scoring chunk per entry (not just first-seen).
  const bestScores = new Map<string, number>();

  for (const { chunk, embedding } of entries) {
    const similarity = cosineSimilarity(queryVector, embedding);
    if (similarity > 0.3) {
      const prev = bestScores.get(chunk.entryId);
      if (prev === undefined || similarity > prev) bestScores.set(chunk.entryId, similarity);
    }
  }

  return [...bestScores.entries()]
    .map(([entryID, score]) => ({ entryID, score }))
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);
}

export function embeddingStats(): { indexed: number; available: boolean } {
  const available = ensureHelper();
  return { indexed: embeddedCount(), available };
}
